### PR TITLE
[gun_adapter] Add proxy creds if passed as proxy_auth in opts

### PR DESCRIPTION
**Pull request by matijavibe**
As GUN already supports proxy with Basic auth, this PR intends to enable passing of auth credentials down to GUN.
Also it adds processing of two known and possible response codes from a proxy server 403 and 407.

### CODE REVIEW

**Comment by teamon** [comment]
@alex-strizhakov thoughts?

File: lib/tesla/adapter/gun.ex [hunk 1]
```diff
@@ -232,17 +232,22 @@ if Code.ensure_loaded?(:gun) do
       end
     end
 
-    defp do_open_conn(uri, %{proxy: {proxy_host, proxy_port}}, gun_opts, tls_opts) do
+    defp do_open_conn(uri, %{proxy: {proxy_host, proxy_port}} = opts, gun_opts, tls_opts) do
       connect_opts =
         uri
         |> tunnel_opts()
         |> tunnel_tls_opts(uri.scheme, tls_opts)
+        |> add_proxy_auth_credentials(opts)
```
Comment (alex-strizhakov) [inline lib/tesla/adapter/gun.ex:240]: Would be great to have auth for `socks` proxy too. [Docs](https://ninenines.eu/docs/en/gun/2.0/manual/gun/) -> socks_opts

**Comment by matijavibe** [comment]
> Hey @matijavibe , how's going?

Hey, sorry I just haven't managed to make time for the PR
I agree with the suggestions, and will try to add socks proxy as well, if I find the means to test it.
Will keep you posted

**Comment by matijavibe** [comment]
@teamon hey, do you mind if I open a new PR, that will have both regular and socks proxy, as I've had issues fixing conflicts and rebasing on this branch, and the new one will be from the latest master?

**Comment by matijavibe** [comment]
@teamon @alex-strizhakov I've opened a new PR form the latest master, that has socks and regular proxy auth suport, as I've had some issues with testing from this branch, I hope you don't mind, and we can close this one?

[424](https://github.com/teamon/tesla/pull/424)

**Pull request closed without merging**